File: unlockfps_clr/SetupForm.cpp
#include "SetupForm.h"
#include "Managed.h"

using namespace System::Threading;
using namespace System::Windows::Forms;

namespace unlockfpsclr
{

    Void SetupForm::OnFormClosing(Object^ sender, FormClosingEventArgs^ e)
    {
        if (String::IsNullOrWhiteSpace(settings->GamePath))
            Application::Exit();
    }

    Void SetupForm::btnBrowse_Click(Object^ sender, EventArgs^ e)
    {
        auto dialog = gcnew OpenFileDialog();
        dialog->Filter = "Process (*.exe)|*.exe|All files (*.*)|*.*";
        dialog->FilterIndex = 0;
        dialog->RestoreDirectory = true;

        if (dialog->ShowDialog() == System::Windows::Forms::DialogResult::OK)
        {
            auto selectedFile = dialog->FileName;
            auto directory = Path::GetDirectoryName(selectedFile);
            
            if (!selectedFile->Contains("GenshinImpact.exe") && !selectedFile->Contains("YuanShen.exe"))
            {
                MessageBox::Show("Please select the game exe\nGenshinImpact.exe for OS version\nor\nYuanShen.exe for CN version", "Error", MessageBoxButtons::OK, MessageBoxIcon::Error);
                return;
            }

            if (!File::Exists(directory + "\\UnityPlayer.dll"))
            {
                MessageBox::Show("That's not the right place", "Error", MessageBoxButtons::OK, MessageBoxIcon::Error);
                return;
            }

            if (backgroundWorker)
                backgroundWorker->CancelAsync();
            settings->GamePath = selectedFile;
            settings->Save();
            this->Close();
        }

    }

    Void SetupForm::OnProcessFound(String^ processPath)
    {
        settings->GamePath = processPath;
        settings->Save();
        this->Close();
    }

    Void SetupForm::OnDoWork(Object^ sender, DoWorkEventArgs^ e)
    {
        auto worker = safe_cast<BackgroundWorker^>(sender);
        while (!worker->CancellationPending)
        {
            auto path = Managed::TryGetGamePath();
            if (!String::IsNullOrWhiteSpace(path))
            {
                settings->GamePath = path;
                settings->Save();
                worker->ReportProgress(100);
                break;
            }

            Thread::Sleep(200);
        }
        
    }

    Void SetupForm::OnProgressChanged(Object^ sender, ProgressChangedEventArgs^ e)
    {
        auto object = e->UserState;
        auto percentage = e->ProgressPercentage;
        if (percentage == 100)
            this->Close();
    }

    Void SetupForm::OnLoad(Object^ sender, EventArgs^ e)
    {
        auto result = Managed::TryResolveGamePath();
        if (result->Count)
        {
            btnConfirm->Visible = true;
            //btnBrowse->Visible = false;
            labelResult->Text = String::Format("Found {0} installation of the game", result->Count);
            labelResult->ForeColor = Color::Green;
            labelSelectInstance->Text = "Select the instance you want to use:";
            comboBoxSelectInst->Visible = true;
            comboBoxSelectInst->Items->AddRange(result->ToArray());
            comboBoxSelectInst->SelectedIndex = 0;
            if (!String::IsNullOrWhiteSpace(settings->GamePath))
                comboBoxSelectInst->SelectedIndex = result->IndexOf(settings->GamePath);
        }
        else
        {
            btnConfirm->Visible = false;
            //btnBrowse->Visible = true;
            labelResult->Text = "cannot find your game";
            labelResult->ForeColor = Color::Red;
            labelSelectInstance->Text = "You can open the game now or browse";
            comboBoxSelectInst->Visible = false;

            backgroundWorker = gcnew BackgroundWorker();
            backgroundWorker->WorkerReportsProgress = true;
            backgroundWorker->WorkerSupportsCancellation = true;
            backgroundWorker->DoWork += gcnew DoWorkEventHandler(this, &SetupForm::OnDoWork);
            backgroundWorker->ProgressChanged += gcnew ProgressChangedEventHandler(this, &SetupForm::OnProgressChanged);
            backgroundWorker->RunWorkerAsync();
        }
    }

    Void SetupForm::btnConfirm_Click(Object^ sender, EventArgs^ e)
    {
        settings->GamePath = safe_cast<String^>(comboBoxSelectInst->Items[comboBoxSelectInst->SelectedIndex]);
        settings->Save();
        this->Close();
    }

}

File: unlockfps_nc/SetupForm.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Runtime.InteropServices;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using Microsoft.Win32;
using unlockfps_nc.Model;
using unlockfps_nc.Service;
using unlockfps_nc.Utility;

namespace unlockfps_nc
{
    public partial class SetupForm : Form
    {
        private CancellationTokenSource _cts;

        private readonly ConfigService _configService;
        private readonly Config _config;

        public SetupForm(ConfigService configService)
        {
            InitializeComponent();
            _configService = configService;
            _config = _configService.Config;
        }

        private void SetupForm_Load(object sender, EventArgs e)
        {
            _cts = new();
            Task.Run(PollProcess, _cts.Token);

            LabelCurrentPath.Text = $@"Current Path: {_config.GamePath}";
            LabelResult.Text = @"Searching...";
            LabelResult.ForeColor = Color.Orange;
            Task.Run(SearchGamePath, _cts.Token);
        }

        private void SetupForm_FormClosing(object sender, FormClosingEventArgs e)
        {
            _cts.Cancel();
            _configService.Save();
        }

        private async Task PollProcess()
        {
            // System.Diagnostics.Process will throw access denied
            // use native win32 api instead

            while (!_cts.Token.IsCancellationRequested)
            {
                await Task.Delay(1000);
                IntPtr windowHandle = IntPtr.Zero;
                IntPtr processHandle = IntPtr.Zero;
                string processPath = string.Empty;

                Native.EnumWindows((hWnd, lParam) =>
                {
                    const int maxCount = 256;
                    StringBuilder sb = new StringBuilder(maxCount);

                    Native.GetClassName(hWnd, sb, maxCount);
                    if (sb.ToString() == "UnityWndClass")
                    {
                        Native.GetWindowThreadProcessId(hWnd, out var pid);
                        processHandle = Native.OpenProcess(
                            ProcessAccess.QUERY_LIMITED_INFORMATION |
                            ProcessAccess.TERMINATE |
                            StandardAccess.SYNCHRONIZE, false, pid);
                        var foundPath = ProcessUtils.GetProcessPath(processHandle);
                        if (!foundPath.Contains("YuanShen.exe") && !foundPath.Contains("GenshinImpact.exe"))
                            return true;

                        windowHandle = hWnd;
                        processPath = foundPath;
                        return false;
                    }

                    return true;
                }, IntPtr.Zero);

                if (windowHandle == IntPtr.Zero)
                    continue;

                Native.TerminateProcess(processHandle, 0);
                Native.CloseHandle(processHandle);

                if (string.IsNullOrEmpty(processPath))
                {
                    MessageBox.Show(@$"Failed to find process path{Environment.NewLine}Please use ""Browse"" instead", @"Error",
                        MessageBoxButtons.OK, MessageBoxIcon.Error);
                    return;
                }

                MessageBox.Show($@"Game Found!{Environment.NewLine}{processPath}", @"Success", MessageBoxButtons.OK,
                    MessageBoxIcon.Information);

                _config.GamePath = processPath;
                Invoke(Close);
            }

        }

        private void SearchGamePath()
        {
            using var hypGlobal = Registry.CurrentUser.OpenSubKey(@"SOFTWARE\Cognosphere\HYP\1_1\hk4e_global");
            using var hypCn = Registry.CurrentUser.OpenSubKey(@"SOFTWARE\miHoYo\HYP\1_1\hk4e_cn");

            List<string> gamePaths = new();

            var installPathGlobal = hypGlobal?.GetValue("GameInstallPath") as string;
            var installPathCn = hypCn?.GetValue("GameInstallPath") as string;

            if (installPathGlobal != null) {
                var path = Path.Combine(installPathGlobal, "GenshinImpact.exe");
                gamePaths.Add(path.Replace('/', '\\'));
            }

            if (installPathCn != null) {
                var path = Path.Combine(installPathCn, "YuanShen.exe");
                gamePaths.Add(path.Replace('/', '\\'));
            }

            Invoke(() =>
            {
                LabelResult.ForeColor = gamePaths.Count > 0 ? Color.Green : Color.Red;
                LabelResult.Text = $@"Found {gamePaths.Count} installation of the game";
                ComboResult.Items.AddRange(gamePaths.ToArray());
                if (gamePaths.Count > 0)
                    ComboResult.SelectedIndex = 0;
            });
        }

        private void BtnBrowse_Click(object sender, EventArgs e)
        {
            if (BrowseDialog.ShowDialog() != DialogResult.OK)
                return;

            var selectedFile = BrowseDialog.FileName;
            var fileName = Path.GetFileNameWithoutExtension(selectedFile);
            var directory = Path.GetDirectoryName(selectedFile);

            if (fileName != "GenshinImpact" && fileName != "YuanShen")
            {
                MessageBox.Show(
                    $@"Please select the game exe{Environment.NewLine}GenshinImpact.exe or YuanShen.exe",
                    @"Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
                return;
            }

            var dataDir = Path.Combine(directory, $"{fileName}_Data");
            if (!Directory.Exists(dataDir))
            {
                MessageBox.Show(@"That's not the right place", @"Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
                return;
            }

            _config.GamePath = selectedFile;
            Close();
        }

        private void BtnConfirm_Click(object sender, EventArgs e)
        {
            var selectedPath = (string)ComboResult.SelectedItem;
            if (string.IsNullOrEmpty(selectedPath))
                return;

            _config.GamePath = selectedPath;
            Close();
        }
    }
}
